### PR TITLE
8264718: Shenandoah: enable string deduplication during root scanning

### PR DESCRIPTION
**Pull request by zhengyu123**
Shenandoah used to scan roots at pauses, so it deliberately disables string deduplication during root scanning to avoid extra pause times.

Now, Shenandoah scans roots in concurrent phase, it is no longer a concern, we should enable it.

Test:
- [x] hotspot_gc_shenandoah

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264718](https://bugs.openjdk.java.net/browse/JDK-8264718): Shenandoah: enable string deduplication during root scanning

### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to d3d63b3808de72007ad92cff4c55d4efe555144e

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3348/head:pull/3348` \
`$ git checkout pull/3348`

Update a local copy of the PR: \
`$ git checkout pull/3348` \
`$ git pull https://git.openjdk.java.net/jdk pull/3348/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3348`

View PR using the GUI difftool: \
`$ git pr show -t 3348`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3348.diff">https://git.openjdk.java.net/jdk/pull/3348.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahSTWMark.cpp [hunk 1]
```diff
@@ -99,8 +128,14 @@ void ShenandoahSTWMark::mark() {
 }
 
 void ShenandoahSTWMark::mark_roots(uint worker_id) {
-  ShenandoahInitMarkRootsClosure  init_mark(task_queues()->queue(worker_id));
-  _root_scanner.roots_do(&init_mark, worker_id);
+  if (ShenandoahStringDedup::is_enabled()) {
+    ShenandoahInitMarkRootsClosure<ENQUEUE_DEDUP>  init_mark(task_queues()->queue(worker_id));
+    _root_scanner.roots_do(&init_mark, worker_id);
+
```
Comment (casparcwang) [inline src/hotspot/share/gc/shenandoah/shenandoahSTWMark.cpp:134]: extra empty line

File: src/hotspot/share/gc/shenandoah/shenandoahSTWMark.cpp [hunk 2]
```diff
@@ -99,8 +128,13 @@ void ShenandoahSTWMark::mark() {
 }
 
 void ShenandoahSTWMark::mark_roots(uint worker_id) {
-  ShenandoahInitMarkRootsClosure  init_mark(task_queues()->queue(worker_id));
-  _root_scanner.roots_do(&init_mark, worker_id);
+  if (ShenandoahStringDedup::is_enabled()) {
+    ShenandoahInitMarkRootsClosure<ENQUEUE_DEDUP>  init_mark(task_queues()->queue(worker_id));
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahSTWMark.cpp:132]: Excess space here and on the other branch, between `<ENQUEUE_DEDUP>` and `init_mark`.

**Review by shipilev — APPROVED** [review]
Looks fine, with a minor nit. Fix it before push.

**Pull request closed without merging**